File: frontend/src/types/copilot.ts
import type { Component } from 'vue'

export type CopilotMode = 'ask' | 'guide' | 'act'
export type CopilotMessageRole = 'user' | 'assistant' | 'system'
export type CopilotCardKind = 'actions' | 'data' | 'task' | 'draft-preview' | 'confirmation' | 'task-plan' | 'task-progress' | 'task-recovery' | 'test-result' | 'preflight' | 'computer-use-consent' | 'computer-use-session'

export interface CopilotToolExecution {
  tool: string
  input: Record<string, unknown>
  preview?: Record<string, unknown>
}

export interface CopilotPageAction {
  id: string
  label: string
  command:
    | { type: 'navigate'; path: string }
    | { type: 'highlight'; targetId: string; label?: string }
    | { type: 'scroll_to'; targetId: string; label?: string }
    | { type: 'switch_tab'; tab: string }
    | { type: 'open_dialog'; targetId: string; label?: string }
    | { type: 'prefill_form'; targetId: string; values: Record<string, unknown> }
    | { type: 'focus_field'; targetId: string; field: string }
}

export interface CopilotPageContext {
  routeName: string
  pageTitle: string
  description?: string
  workspaceId?: string
  selectedEntity?: {
    type: string
    id: string
    name?: string
  }
  availableActions: CopilotPageAction[]
  suggestedPrompts: string[]
}

export interface CopilotAction {
  label: string
  variant?: 'primary' | 'secondary'
  command?: CopilotPageAction['command']
  event?: 'confirm_create_draft' | 'confirm_execute_tool' | 'confirm_task_plan' | 'pause_task' | 'resume_task' | 'cancel_task' | 'retry_task' | 'manual_create_agent' | 'cancel_operation' | 'retry' | 'confirm_computer_use' | 'stop_computer_use'
}

export interface CopilotCard {
  kind: CopilotCardKind
  title: string
  description?: string
  items?: Record<string, unknown>[]
  actions?: CopilotAction[]
  payload?: Record<string, unknown>
}

export interface CopilotMessage {
  id: string
  role: CopilotMessageRole
  content: string
  createdAt: string
  cards?: CopilotCard[]
}

export interface CopilotTool {
  name: string
  description: string
  category: string
  riskLevel: 'L0' | 'L1' | 'L2' | 'L3'
  requiresConfirmation: boolean
  permission: string[]
  inputSchema?: Record<string, unknown>
  readOnly?: boolean
}

export interface CopilotResponse {
  message: string
  cards?: CopilotCard[]
}

export interface CopilotQuickPrompt {
  label: string
  prompt: string
  icon?: Component
}

File: backend/app/schemas/copilot.py
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class CopilotToolOut(BaseModel):
    name: str
    description: str
    category: str
    risk_level: Literal["L0", "L1", "L2", "L3"]
    requires_confirmation: bool
    permission: list[str]
    input_schema: dict[str, Any]


class CopilotExecuteIn(BaseModel):
    model_config = {"extra": "forbid"}

    tool: str = Field(min_length=1)
    input: dict[str, Any] = Field(default_factory=dict)
    confirmed: bool = False
    conversation_id: str | None = None
    workspace_id: str | None = None
    preview: dict[str, Any] | None = None


class CopilotExecuteOut(BaseModel):
    tool: str
    status: Literal["success", "failed"]
    message: str
    target_type: str | None = None
    target_id: UUID | None = None
    result: dict[str, Any] = Field(default_factory=dict)
    audit: dict[str, Any] = Field(default_factory=dict)
